hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateManagerV2Impl.java

```diff
@@ -0,0 +1,233 @@
+      addPipeline(pipeline.getProtobufMessage());
+    }
+  }
+
+  @Override
+  public void addPipeline(HddsProtos.Pipeline pipelineProto)
+      throws IOException {
+    Pipeline pipeline = Pipeline.getFromProtobuf(pipelineProto);
+    if (pipelineStore != null) {
+      pipelineStore.put(pipeline.getId(), pipeline);
+    }
+    pipelineStateMap.addPipeline(pipeline);
+    if (nodeManager != null) {
+      nodeManager.addPipeline(pipeline);
+    }
+    LOG.info("Created pipeline {}.", pipeline);
+  }
+
+  @Override
+  public void addContainerToPipeline(
+      PipelineID pipelineId, ContainerID containerID)
+      throws IOException {
+    pipelineStateMap.addContainerToPipeline(pipelineId, containerID);
+  }
+
+  @Override
+  public Pipeline getPipeline(PipelineID pipelineID)
+      throws PipelineNotFoundException {
+    return pipelineStateMap.getPipeline(pipelineID);
+  }
+
+  @Override
+  public List<Pipeline> getPipelines() {
+    return pipelineStateMap.getPipelines();
+  }
+
+  @Override
+  public List<Pipeline> getPipelines(HddsProtos.ReplicationType type) {
+    return pipelineStateMap.getPipelines(type);
+  }
+
+  @Override
+  public List<Pipeline> getPipelines(
+      HddsProtos.ReplicationType type, HddsProtos.ReplicationFactor factor) {
+    return pipelineStateMap.getPipelines(type, factor);
+  }
+
+  @Override
+  public List<Pipeline> getPipelines(
+      HddsProtos.ReplicationType type, HddsProtos.ReplicationFactor factor,
+                              Pipeline.PipelineState state) {
+    return pipelineStateMap.getPipelines(type, factor, state);
+  }
+
+  @Override
+  public List<Pipeline> getPipelines(
+      HddsProtos.ReplicationType type, HddsProtos.ReplicationFactor factor,
+      Pipeline.PipelineState state, Collection<DatanodeDetails> excludeDns,
+      Collection<PipelineID> excludePipelines) {
+    return pipelineStateMap
+        .getPipelines(type, factor, state, excludeDns, excludePipelines);
+  }
+
+  @Override
+  public List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                                     Pipeline.PipelineState... states) {
+    return pipelineStateMap.getPipelines(type, states);
+  }
+
+  @Override
+  public NavigableSet<ContainerID> getContainers(PipelineID pipelineID)
+      throws IOException {
+    return pipelineStateMap.getContainers(pipelineID);
+  }
+
+  @Override
+  public int getNumberOfContainers(PipelineID pipelineID) throws IOException {
+    return pipelineStateMap.getNumberOfContainers(pipelineID);
+  }
+
+  @Override
+  public void removePipeline(HddsProtos.PipelineID pipelineIDProto)
+      throws IOException {
+    PipelineID pipelineID = PipelineID.getFromProtobuf(pipelineIDProto);
+    if (pipelineStore != null) {
+      pipelineStore.delete(pipelineID);
+    }
+    Pipeline pipeline = pipelineStateMap.removePipeline(pipelineID);
+    if (nodeManager != null) {
+      nodeManager.removePipeline(pipeline);
+    }
```

Comment:
We should not do `null` check here. If `pipelineStore` and `nodeManager` are `null` there is a problem/bug.